File: Assets/src/model/enemies/Butterfly.cs
using System.Collections.Generic;
using System.Linq;
using UnityEngine;

[System.Serializable]
[ObjectInfo("butterfly", flavorText: "Delicate and gentle, this Butterfly seems drawn to you, waiting for your command.")]
public class ItemButterfly : Item, IUsable {
  public void Use(Actor a) {
    var tile = Util.RandomPick(a.floor.GetAdjacentTiles(a.pos).Where((t) => t.CanBeOccupied()));
    if (tile != null) {
      a.floor.Put(new Butterfly(tile.pos));
      Destroy();
    }
  }

  internal override string GetStats() => "Summons an allied Butterfly. Every 5 turns, the Butterfly duplicates the Grass you're standing on to the four cardinally adjacent tiles.";
}

[System.Serializable]
[ObjectInfo(description: "Every 5 turns, the Butterfly duplicates the Grass you're standing on to the four cardinally adjacent tiles.")]
public class Butterfly : AIActor {

  private static float DUPLICATE_CD = 5;
  float cooldown = 0;

  public Butterfly(Vector2Int pos) : base(pos) {
    faction = Faction.Ally;
    hp = baseMaxHp = 1;
    ClearTasks();
    statuses.Add(new CharmedStatus());
  }

  protected override ActorTask GetNextTask() {
    var player = GameModel.main.player;

    // we want to duplicate
    if (cooldown <= 0 && player.grass != null) {
      // we're on top of a grass, duplicate!
      return new TelegraphedTask(this, 1, new GenericBaseAction(this, DuplicateGrass));
    }
    // we can't duplicate, either because there's nothing nearby or we're on CD
    if (cooldown > 0) {
      cooldown--;
    }
    return new ChaseTargetTask(this, player);
  }

  private void DuplicateGrass() {
    var grass = GameModel.main.player.grass;
    if (grass != null) {
      var neighborTiles = grass.floor.GetCardinalNeighbors(grass.pos).Where((tile) => tile is Ground && tile.grass == null);
      var constructorInfo = grass.GetType().GetConstructor(new System.Type[1] { typeof(Vector2Int) });
      foreach (var tile in neighborTiles) {
        var newGrass = (Grass)constructorInfo.Invoke(new object[] { tile.pos });
        grass.floor.Put(newGrass);
      }
      cooldown = DUPLICATE_CD;
    }
  }
}

File: web/src/model/enemies/Butterfly.ts
import { AIActor } from './AIActor';
import { ActorTask } from '../ActorTask';
import { ChaseTargetTask } from '../tasks/ChaseTargetTask';
import { TelegraphedTask } from '../tasks/TelegraphedTask';
import { GenericBaseAction } from '../BaseAction';
import { Vector2Int } from '../../core/Vector2Int';
import { Faction } from '../../core/types';
import { GameModelRef } from '../GameModelRef';
import { Grass } from '../grasses/Grass';
import { Ground } from '../Tile';
import { CharmedStatus } from '../statuses/CharmedStatus';
import { entityRegistry } from '../../generator/entityRegistry';

/**
 * Allied butterfly. Every 5 turns, duplicates the player's grass to cardinal neighbors.
 * Port of C# Butterfly.cs.
 */
export class Butterfly extends AIActor {
  private static DUPLICATE_CD = 5;
  private cooldown = 0;

  constructor(pos: Vector2Int) {
    super(pos);
    this.faction = Faction.Ally;
    this._hp = this._baseMaxHp = 1;
    this.clearTasks();
    this.statuses.add(new CharmedStatus());
  }

  baseAttackDamage(): [number, number] {
    return [0, 0];
  }

  protected getNextTask(): ActorTask {
    const player = GameModelRef.main.player;

    // We want to duplicate if off cooldown and player is on grass
    if (this.cooldown <= 0 && player.grass != null) {
      return new TelegraphedTask(this, 1, new GenericBaseAction(this, () => this.duplicateGrass()));
    }

    // Can't duplicate — either on cooldown or no grass
    if (this.cooldown > 0) {
      this.cooldown--;
    }
    return new ChaseTargetTask(this, player);
  }

  private duplicateGrass(): void {
    const player = GameModelRef.main.player;
    const grass = player.grass;
    if (grass == null || !(grass instanceof Grass)) return;

    const floor = grass.floor!;
    const neighbors = floor.getCardinalNeighbors(grass.pos)
      .filter(tile => tile instanceof Ground && floor.grasses.get(tile.pos) == null);

    // Use the grass constructor to create copies
    const GrassType = grass.constructor as new (pos: Vector2Int) => Grass;
    for (const tile of neighbors) {
      try {
        const newGrass = new GrassType(tile.pos);
        floor.put(newGrass);
      } catch {
        // Constructor may not match expected signature — skip
      }
    }
    this.cooldown = Butterfly.DUPLICATE_CD;
  }
}

entityRegistry.register('Butterfly', Butterfly);
